### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by denizozger** [cayasso/primus-rooms#25]
I have a very simple application that uses primus-rooms (https://github.com/denizozger/node-primus-server/blob/master/server.js). The application works well for ~10minues, then memory seems to increase infinitely without nothing changing with regards to connections. More detail can be found here (https://medium.com/node-js-javascript/b63bfca0539) under "Primus" title. 

Any ideas?

Thanks

**Comment by devinivy** [comment]
I'd like to dig this back up.  Will take a look, would love some help.

**Comment by cayasso** [comment]
Hey @lpinca what do you think about this issue, Its been a long time now, should we leave it like that and close or should we pursue a fix for the "removed reference"? I like your idea of the check, we should do this in the less code possible.

**Comment by lpinca** [comment]
I would close this, no one reported the above issue which means it's something uncommon.

**Comment by cayasso** [comment]
Ok closing...

**Comment by lpinca** [comment]
@cayasso I reopen the issue as the leak is not fixed. Here is a test case:
```js
'use strict';

const Primus = require('primus');
const http = require('http');

const server = http.createServer();
const primus = new Primus(server);

primus.use('rooms', 'primus-rooms');

primus.on('connection', (spark) => {
  console.log('%s connected', spark.id);

  spark.end();

  process.nextTick(() => {
    //
    // Assume this is the result of an async operation that completes right
    // after the spark is closed.
    //
    spark.join('foo', () => {
      console.log(primus.adapter.rooms);
      console.log(primus.adapter.sids);
    });
  });
});
primus.on('disconnection', (spark) => {
  console.log('%s disconnected', spark.id);
});
primus.on('leaveallrooms', function (rooms, spark) {
  console.log('%s left rooms', spark.id, rooms);
});

server.listen(function () {
  new primus.Socket('http://localhost:' + server.address().port);
});
```

**Comment by devinivy** [comment]
I can't officially take this issue over, but I will look into it when I get a chance!

**Comment by lpinca** [comment]
I've just found another leak in the adapter. I'll start with fixing it first, probably later today.

**Comment by lpinca** [comment]
Something like this should work:
```diff
diff --git a/lib/rooms.js b/lib/rooms.js
index 9c776c3..e01c70f 100644
--- a/lib/rooms.js
+++ b/lib/rooms.js
@@ -133,11 +133,15 @@ Rooms.prototype.join = function join(room, fn) {
 Rooms.prototype._join = function _join(room, fn) {
   var rm = this;
   debug('joining room %s', room);
-  this.adapter.add(this.id, room, function set(err, res) {
-    if (err) return fn(new RoomsError(err, rm.ctx));
-    rm.emits('joinroom', room);
-    fn(null, res);
-  });
+  if (this.connections[this.id]) {
+    this.adapter.add(this.id, room, function set(err, res) {
+      if (err) return fn(new RoomsError(err, rm.ctx));
+      rm.emits('joinroom', room);
+      fn(null, res);
+    });
+  } else {
+    fn(new RoomsError('Spark is closed', this.ctx));
+  }
   return this;
 };
```
but I wonder if the callback should be called in the next tick to keep it async.
It may have side effects when joining multiple rooms at once or when using multiple sparks at once.
Also other issues may arise because references are removed as stated above.

I'm no affected by this issue because I only join spaks individually anche I check if they are closed before calling `join`.

**Comment by devinivy** [comment]
Yes, I think it should definitely be called with `setImmediate()`.

**Comment by lpinca** [comment]
What about this:

- Do not delete the references when the spark closes.
- Add the above hack.

This is probably going too far but I wonder if it makes sense to:

- Rewrite the memory adapter to only allow an async interface. Any real (db based) adapter would only use an async interface.
- Update primus-rooms accordingly.

**Comment by cayasso** [comment]
@lpinca I like both, I think we can do both, yesterday I was thinking on rewriting the adapter and make it more ES6 friendly and use setImmediate instead of process.nextTick to avoid the binding jazz and make stuff easier, and also to make it a function instead of a class which I find to be overhead these days. The point of making it async will bring consistency to the adapter. 

I will throw some code in a 1.0 branch and @lpica maybe (if you have some time) if you can help with the first two points you mentioned as a solution for 0.x versions

What do you guys think?

**Comment by lpinca** [comment]
@cayasso sounds good. I was replacing `process.nextTick` with `setImmediate` just now in the memory adapter :smile:.

**Comment by cayasso** [comment]
@lpinca awesome go ahead! ;-)
